disttask/loaddata/dispatcher.go

```diff
@@ -17,36 +17,126 @@ package loaddata
 	"github.com/pingcap/tidb/executor/asyncloaddata"
 	"github.com/pingcap/tidb/executor/importer"
 	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/etcd"
 	"github.com/pingcap/tidb/util/logutil"
 	"go.uber.org/atomic"
 	"go.uber.org/zap"
 )
 
+const (
+	registerTaskTTL        = 10 * time.Minute
+	refreshTaskTTLInterval = 3 * time.Minute
+	registerTimeout        = 5 * time.Second
+)
+
+// NewTaskRegisterWithTTL is the ctor for TaskRegister.
+// It is exported for testing.
+var NewTaskRegisterWithTTL = utils.NewTaskRegisterWithTTL
+
+type taskInfo struct {
+	taskID int64
+
+	sync.RWMutex
+	lastRegisterTime atomic.Time
+
+	// initialized lazily in register()
+	etcdClient   *etcd.Client
+	taskRegister utils.TaskRegister
+}
+
+func (t *taskInfo) register(ctx context.Context) {
+	if time.Since(t.lastRegisterTime.Load()) < refreshTaskTTLInterval {
+		return
+	}
+
+	t.Lock()
+	defer t.Unlock()
+	if time.Since(t.lastRegisterTime.Load()) < refreshTaskTTLInterval {
+		return
+	}
+	logger := logutil.BgLogger().With(zap.Int64("task_id", t.taskID))
+	if t.taskRegister == nil {
+		client, err := importer.GetEtcdClient()
+		if err != nil {
+			logger.Warn("get etcd client failed", zap.Error(err))
+			return
+		}
+		t.etcdClient = client
+		t.taskRegister = NewTaskRegisterWithTTL(client.GetClient(), registerTaskTTL,
+			utils.RegisterImportInto, strconv.FormatInt(t.taskID, 10))
+	}
+	timeoutCtx, cancel := context.WithTimeout(ctx, registerTimeout)
+	defer cancel()
+	if err := t.taskRegister.RegisterTaskOnce(timeoutCtx); err != nil {
+		logger.Warn("register task failed", zap.Error(err))
+	} else {
+		logger.Info("register task to pd or refresh lease success")
+	}
+	// we set it even if register failed, TTL is 10min, refresh interval is 3min,
+	// we can try 2 times before the lease is expired.
+	t.lastRegisterTime.Store(time.Now())
+}
+
+func (t *taskInfo) close(ctx context.Context) {
+	t.Lock()
+	defer t.Unlock()
+	logger := logutil.BgLogger().With(zap.Int64("task_id", t.taskID))
+	if t.taskRegister != nil {
+		timeoutCtx, cancel := context.WithTimeout(ctx, registerTimeout)
+		defer cancel()
+		if err := t.taskRegister.Close(timeoutCtx); err != nil {
+			logger.Warn("unregister task failed", zap.Error(err))
+		} else {
+			logger.Info("unregister task success")
+		}
+		t.taskRegister = nil
+	}
+	if t.etcdClient != nil {
+		if err := t.etcdClient.Close(); err != nil {
+			logger.Warn("close etcd client failed", zap.Error(err))
+		}
+		t.etcdClient = nil
+	}
+}
+
 type flowHandle struct {
 	mu sync.RWMutex
 	// the last time we switch TiKV into IMPORT mode, this is a global operation, do it for one task makes
 	// no difference to do it for all tasks. So we do not need to record the switch time for each task.
 	lastSwitchTime atomic.Time
+
+	// taskInfoMap is a map from taskID to taskInfo
+	taskInfoMap sync.Map
 }
 
 var _ dispatcher.TaskFlowHandle = (*flowHandle)(nil)
 
 func (h *flowHandle) OnTicker(ctx context.Context, task *proto.Task) {
```

Comment:
Will this callback function be called concurrently by framework 🤔 Maybe no need to use sync.Map

Reply:
i am considering to remove all sync, since we only allow 1 load task right now.